File: CSharp/SpecStudioExamplesTestsCSharp/common/FrameDisplayTyped.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using System;
using System.Buffers;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using System.Text.Json;
using production;

    public class FrameDisplayTyped
    {
        public string frame;
        public string mark1;
        public string mark2;
        public string mark3;
        public string totalScore;

        public FrameDisplayTyped(string frame, string mark1, string mark2, string mark3, string totalScore)
        {
            this.frame = frame;
            this.mark1 = mark1;
            this.mark2 = mark2;
            this.mark3 = mark3;
            this.totalScore = totalScore;
        }

        public FrameDisplayString ToFrameDisplayString()
        {
            return new FrameDisplayString(
                Json.ToText(this.frame),
                Json.ToText(this.mark1),
                Json.ToText(this.mark2),
                Json.ToText(this.mark3),
                Json.ToText(this.totalScore)
            );
        }

        public static List<FrameDisplayString> ToStringList(List<FrameDisplayTyped> list)
        {
            var result = new List<FrameDisplayString>();
            foreach (var t in list) result.Add(t.ToFrameDisplayString());
            return result;
        }

        public static List<FrameDisplayTyped> FromStringList(List<FrameDisplayString> list)
        {
            var result = new List<FrameDisplayTyped>();
            foreach (var s in list) result.Add(s.ToFrameDisplayTyped());
            return result;
        }

        public void WriteJson(Utf8JsonWriter w)
        {
            w.WriteStartObject();
            w.WriteString("frame", this.frame);
            w.WriteString("mark1", this.mark1);
            w.WriteString("mark2", this.mark2);
            w.WriteString("mark3", this.mark3);
            w.WriteString("totalScore", this.totalScore);
            w.WriteEndObject();
        }

        public string ToJSON()
        {
            var buffer = new ArrayBufferWriter<byte>();
            using (var w = new Utf8JsonWriter(buffer)) { WriteJson(w); }
            return Encoding.UTF8.GetString(buffer.WrittenSpan);
        }

        public static FrameDisplayTyped FromJsonElement(JsonElement m)
        {
            return new FrameDisplayTyped(
                Json.AsString(Json.Require(m, "frame"), "frame"),
                Json.AsString(Json.Require(m, "mark1"), "mark1"),
                Json.AsString(Json.Require(m, "mark2"), "mark2"),
                Json.AsString(Json.Require(m, "mark3"), "mark3"),
                Json.AsString(Json.Require(m, "totalScore"), "totalScore")
            );
        }

        public static FrameDisplayTyped FromJSON(string json)
        {
            return FromJsonElement(Json.Parse(json));
        }

        public static string ToJSONList(List<FrameDisplayTyped> list)
        {
            var buffer = new ArrayBufferWriter<byte>();
            using (var w = new Utf8JsonWriter(buffer))
            {
                w.WriteStartArray();
                foreach (var item in list) item.WriteJson(w);
                w.WriteEndArray();
            }
            return Encoding.UTF8.GetString(buffer.WrittenSpan);
        }

        public static List<FrameDisplayTyped> FromJSONList(string json)
        {
            var result = new List<FrameDisplayTyped>();
            var root = Json.Parse(json);
            Json.RequireArray(root, "FrameDisplayTyped");
            foreach (var e in root.EnumerateArray()) result.Add(FromJsonElement(e));
            return result;
        }

        public override string ToString()
        {
            return $"Frame={frame}, Mark1={mark1}, Mark2={mark2}, Mark3={mark3}, TotalScore={totalScore}";
        }

        public override bool Equals(object? obj)
        {
            if (obj is not FrameDisplayTyped other) return false;
            return object.Equals(this.frame, other.frame)
                && object.Equals(this.mark1, other.mark1)
                && object.Equals(this.mark2, other.mark2)
                && object.Equals(this.mark3, other.mark3)
                && object.Equals(this.totalScore, other.totalScore);
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.frame);
            h.Add(this.mark1);
            h.Add(this.mark2);
            h.Add(this.mark3);
            h.Add(this.totalScore);
            return h.ToHashCode();
        }
    }
}

File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/FrameDisplayTyped.java
package spectable.common;

import java.util.ArrayList;
import java.util.List;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class FrameDisplayTyped {
    public String frame;
    public String mark1;
    public String mark2;
    public String mark3;
    public String totalScore;

    public FrameDisplayTyped(String frame, String mark1, String mark2, String mark3, String totalScore) {
        this.frame = frame;
        this.mark1 = mark1;
        this.mark2 = mark2;
        this.mark3 = mark3;
        this.totalScore = totalScore;
    }

    public FrameDisplayTyped(FrameDisplayString s) {
        this.frame = s.frame;
        this.mark1 = s.mark1;
        this.mark2 = s.mark2;
        this.mark3 = s.mark3;
        this.totalScore = s.totalScore;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof FrameDisplayTyped)) return false;
        FrameDisplayTyped that = (FrameDisplayTyped) o;
        return Objects.equals(frame, that.frame)
            && Objects.equals(mark1, that.mark1)
            && Objects.equals(mark2, that.mark2)
            && Objects.equals(mark3, that.mark3)
            && Objects.equals(totalScore, that.totalScore);
    }

    @Override
    public int hashCode() {
        return Objects.hash(frame, mark1, mark2, mark3, totalScore);
    }

    public FrameDisplayString toFrameDisplayString() {
        return new FrameDisplayString(String.valueOf(frame), String.valueOf(mark1), String.valueOf(mark2), String.valueOf(mark3), String.valueOf(totalScore));
    }

    public static List<FrameDisplayTyped> fromStringList(List<FrameDisplayString> list) {
        List<FrameDisplayTyped> result = new ArrayList<>();
        for (FrameDisplayString s : list) result.add(new FrameDisplayTyped(s));
        return result;
    }

    public static List<FrameDisplayString> toStringList(List<FrameDisplayTyped> list) {
        List<FrameDisplayString> result = new ArrayList<>();
        for (FrameDisplayTyped t : list) result.add(t.toFrameDisplayString());
        return result;
    }

    /** Internal plumbing for {@link Json}; use toJSON() for JSON text. */
    public Map<String, Object> toJsonValue() {
        Map<String, Object> m = new LinkedHashMap<String, Object>();
        m.put("frame", frame);
        m.put("mark1", mark1);
        m.put("mark2", mark2);
        m.put("mark3", mark3);
        m.put("totalScore", totalScore);
        return m;
    }

    public String toJSON() { return Json.write(toJsonValue()); }

    /** Internal plumbing for {@link Json}; use fromJSON(String) for JSON text. */
    public static FrameDisplayTyped fromJsonValue(Map<String, Object> m) {
        if (m == null) return null;
        return new FrameDisplayTyped(
            Json.asString(Json.require(m, "frame"), "frame"),
            Json.asString(Json.require(m, "mark1"), "mark1"),
            Json.asString(Json.require(m, "mark2"), "mark2"),
            Json.asString(Json.require(m, "mark3"), "mark3"),
            Json.asString(Json.require(m, "totalScore"), "totalScore"));
    }

    public static FrameDisplayTyped fromJSON(String json) {
        return fromJsonValue(Json.parseObject(json));
    }

    public static String toJSONList(List<FrameDisplayTyped> list) {
        List<Object> arr = new ArrayList<Object>();
        if (list != null)
            for (FrameDisplayTyped item : list) arr.add(item == null ? null : item.toJsonValue());
        return Json.write(arr);
    }

    public static List<FrameDisplayTyped> fromJSONList(String json) {
        List<FrameDisplayTyped> result = new ArrayList<FrameDisplayTyped>();
        for (Object e : Json.parseArray(json))
            result.add(fromJsonValue(Json.asObject(e, "FrameDisplayTyped")));
        return result;
    }
}
